src/app-layer-ssl.c

```diff
@@ -287,11 +289,46 @@ static int TLSDecodeHandshakeHello(SSLState *ssl_state, uint8_t *input,
     if (!(HAS_SPACE(2)))
         goto invalid_length;
 
-    /* skip cipher suites */
-    uint16_t cipher_suites_length = input[0] << 8 | input[1];
-    input += 2;
+    if (ssl_state->curr_connp->handshake_type == SSLV3_HS_CLIENT_HELLO) {
+        uint16_t cipher_suites_length = input[0] << 8 | input[1];
+        input += 2;
+
+        if (cipher_suites_length & 1) {
+            SCLogDebug("Invalid length for cipher suites");
+            goto invalid_length;
+        }
+
+        if (!HAS_SPACE(cipher_suites_length)) {
+            SCLogDebug("Invalid length for cipher suites");
+            goto invalid_length;
+        }
+
+        ssl_state->curr_connp->num_cipher_suites = 0;
+        ssl_state->curr_connp->cipher_suites = NULL;
+
+        if (cipher_suites_length >= 2) {
+            ssl_state->curr_connp->num_cipher_suites = cipher_suites_length / 2;
+            ssl_state->curr_connp->cipher_suites = SCMalloc(cipher_suites_length);
+            if (!ssl_state->curr_connp->cipher_suites) {
+                SCLogError(SC_ERR_MEM_ALLOC, "Failed to allocate client cipher suites.");
+            } else {
+                memcpy(ssl_state->curr_connp->cipher_suites, input, cipher_suites_length);
+            }
+            input += cipher_suites_length;
+        }
+    }
+    if (ssl_state->curr_connp->handshake_type == SSLV3_HS_SERVER_HELLO) {
+        ssl_state->curr_connp->cipher_suites = SCMalloc(sizeof(uint16_t));
```

Comment:
I've seen a memleak happen here:
<pre>
==2348==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 2 byte(s) in 1 object(s) allocated from:
    #0 0x4c2e9b in __interceptor_malloc (/home/victor/qa/buildbot/donkey/z600fuzz/Private/src/suricata+0x4c2e9b)
    #1 0x837079 in TLSDecodeHandshakeHello /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer-ssl.c:321:48
    #2 0x82fdd3 in SSLv3ParseHandshakeType /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer-ssl.c:471:18
    #3 0x82ce5d in SSLv3ParseHandshakeProtocol /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer-ssl.c:682:14
    #4 0x828282 in SSLv3Decode /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer-ssl.c:1316:22
    #5 0x8201b6 in SSLDecode /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer-ssl.c:1484:30
    #6 0x819cf7 in SSLParseServerRecord /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer-ssl.c:1576:12
    #7 0x7a2463 in AppLayerParserParse /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer-parser.c:1051:13
    #8 0x53887e in TCPProtoDetect /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer.c:421:17
    #9 0x5353b1 in AppLayerHandleTCPData /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/app-layer.c:561:13
    #10 0x117bd20 in ReassembleUpdateAppLayer /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp-reassemble.c:1030:13
    #11 0x1178ec5 in StreamTcpReassembleAppLayer /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp-reassemble.c:1102:12
    #12 0x1188651 in StreamTcpReassembleHandleSegmentUpdateACK /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp-reassemble.c:1649:9
    #13 0x1187ac2 in StreamTcpReassembleHandleSegment /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp-reassemble.c:1688:9
    #14 0x114e4b1 in HandleEstablishedPacketToServer /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp.c:2096:9
    #15 0x10ddc9f in StreamTcpPacketStateEstablished /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp.c:2474:13
    #16 0x10b545e in StreamTcpPacket /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp.c:4534:20
    #17 0x11202c9 in StreamTcp /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/stream-tcp.c:4898:11
    #18 0xde6c01 in FlowWorker /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/flow-worker.c:215:9
    #19 0x11d6014 in TmThreadsSlotVarRun /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/tm-threads.c:128:17
    #20 0x11f63d1 in TmThreadsSlotVar /home/victor/qa/buildbot/donkey/z600fuzz/Private/src/tm-threads.c:585:17
    #21 0x7fac05e94183 in start_thread /build/eglibc-MjiXCM/eglibc-2.19/nptl/pthread_create.c:312
</pre>

This happens in fuzzer generated traffic. I guess there are 2 possible cases:
1. memory is not freed at flow end (seems unlikely)
2. memory allocation is overwritten. This is what I suspect. Can this function be called multiple times if the traffic is weird enough?